### PR TITLE
Added XLA multi-processing support and API.

### PR DESCRIPTION
**Pull request by dlibenzi**
This is the first sketch of the new multi-processing API.
It uses PyTorch multiprocessing since it provides some facilities we might want to use in the future.
As a restriction, it forces `spawn` as multiprocessing mechanism, which forces all the data to be passed to the child, to be rooted at arguments of the function we pass to `xmp.spawn()`.

PTAL to the MNIST example.
I have left the train/test loop functions, but these are no more a requirement with multiprocessing.
Essentially, as long as the child processes "do the same work" (do generate same graphs), there is no need for dedicated loop functions.

The "fork" happen at the very beginning of the main entry point, and from there every child sees one device (local).
We do use the `DataParallel.ParallelLoader` in order to overlap device data load with TPU computations, but it is not required.
The `xm.optimizer_step()` is still required.

This should work even in BigSlice mode, but I only tested it on single host mode ATM.
Even on single host mode, with multiprocessing there is still a mesh service starting at TPU:0 and other local processes connects to that.
This allows single TPU mesh initialization.

I do expect a few iterations over this, but we can get this reviewed now.

### CODE REVIEW

File: test/test_train_mp_mnist.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+      correct += pred.eq(target.view_as(pred)).sum().item()
+      total_samples += data.size()[0]
+
+    accuracy = 100.0 * correct / total_samples
+    test_utils.print_test_update(device, accuracy)
+    return accuracy
+
+  accuracy = 0.0
+  for epoch in range(1, FLAGS.num_epochs + 1):
+    para_loader = dp.ParallelLoader(train_loader, [device])
```
Comment (zcain117) [inline test/test_train_mp_mnist.py]: do these ParallelLoaders need to created every epoch or can they be created up above where we create the train_loader and test_loader?

File: test/test_train_mp_mnist.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+    para_loader = dp.ParallelLoader(test_loader, [device])
+    accuracy = test_loop_fn(para_loader.per_device_loader(device))
+    if FLAGS.metrics_debug:
+      print(torch_xla._XLAC._xla_metrics_report())
+
+  return accuracy
+
+
+def _mp_fn(index, flags):
```
Comment (zcain117) [inline test/test_train_mp_mnist.py]: can these spawned fn's have a return value? We can probably work around it either way, e.g. by logging and regexing later like we now in nightly testing or asserting in the fn like you do here

File: test/test_train_mp_mnist.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+  global FLAGS
+  FLAGS = flags
+  torch.set_default_tensor_type('torch.FloatTensor')
+  accuracy = train_mnist()
+  if FLAGS.tidy and os.path.isdir(FLAGS.datadir):
+    shutil.rmtree(FLAGS.datadir)
+  if accuracy < FLAGS.target_accuracy:
+    print('Accuracy {} is below target {}'.format(accuracy,
+                                                  FLAGS.target_accuracy))
```
Comment (zcain117) [inline test/test_train_mp_mnist.py]: Just so I understand:
the code in _mp_fn would run only once during setup, e.g. a user would see this 'Accuracy {} is below target...' print statement only once using --num_cores=8, but the code in train_mnist() would run once per device?

I guess to me it's not clear at which point in this file the fork occurs

File: test/test_train_mp_mnist.py (hunk 1 above)
Comment (dlibenzi) [reply test/test_train_mp_mnist.py]: They have to, at this time.
Concurrency gets a bit tricky with its threads, if we support rewind functionalities.

File: test/test_train_mp_mnist.py (hunk 2 above)
Comment (dlibenzi) [reply test/test_train_mp_mnist.py]: The mp_fn are really the real main function of the process.
They can exit with different exit status, but we currently do nothing with it.
Good point though. This is needed in order to properly exit with error status.

File: test/test_train_mp_mnist.py (hunk 3 above)
Comment (dlibenzi) [reply test/test_train_mp_mnist.py]: The mp_fn runs per device. Each process has its own device.
If we want to print only from one device, we can use ordinal==0 checks.

File: test/test_train_mp_mnist.py (hunk 2 above)
Comment (dlibenzi) [reply test/test_train_mp_mnist.py]: Actually, `torch.multiprocessing.spawn()` in wait-mode, already raises an exception if one of the child exists with a code different from 0.

File: test/test_train_mp_mnist.py (hunk 1 above)
Comment (jysohn23) [reply test/test_train_mp_mnist.py]: Just curious, but what happens when a user just creates them once and we replay them every epoch?

File: torch_xla_py/xla_env_vars.py [hunk 4]
```diff
@@ -0,0 +1,8 @@
+MP_DEVICE = 'XRT_MULTI_PROCESSING_DEVICE'
+LOCAL_WORKER = 'XRT_LOCAL_WORKER'
+TPU_CONFIG = 'XRT_TPU_CONFIG'
+SERVICE_ADDRESS = 'XRT_MESH_SERVICE_ADDRESS'
+DEVICE_MAP = 'XRT_DEVICE_MAP'
+WORKERS = 'XRT_WORKERS'
+ORDINAL = 'XRT_SHARD_ORDINAL'
```
Comment (jysohn23) [inline torch_xla_py/xla_env_vars.py]: If I'm understanding this PR correctly, this should work just fine with `xla_dist.py`?

`XRT_LOCAL_WORKER` is still per VM, `XRT_SHARD_ORDINAL` is overridden (so no need to set this anymore in `xla_dist.py` unless we're not using MP for distributed training), `WORLD_SIZE` is still the total number of VMs.

File: torch_xla_py/xla_multiprocessing.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+  worker = os.environ.get(xenv.LOCAL_WORKER, None)
+  return int(worker.split(':')[1]) if worker is not None else 0
+
+
+def _local_index_to_global(index):
+  return _get_local_worker_index() * _get_devices_per_worker() + index
+
+
+def _get_world_size():
+  wsize = int(os.environ.get(xenv.WORLD_SIZE, '1'))
```
Comment (jysohn23) [inline torch_xla_py/xla_multiprocessing.py]: We can use `xm.xrt_world_size()` instead?

File: test/test_train_mp_mnist.py (hunk 1 above)
Comment (dlibenzi) [reply test/test_train_mp_mnist.py]: They do not support "rewind". So you get EOF (no more elements) ... I think.

File: torch_xla_py/xla_multiprocessing.py (hunk 5 above)
Comment (dlibenzi) [reply torch_xla_py/xla_multiprocessing.py]: > We can use `xm.xrt_world_size()` instead?

We need to be careful in calling into `xm` because we do not know if calls into that code, triggers XLA library initialization (which they should not at some stages in the multiprocessing code).

File: torch_xla_py/xla_env_vars.py (hunk 4 above)
Comment (dlibenzi) [reply torch_xla_py/xla_env_vars.py]: There should not be any changes in `xla_dist.py`. That module should not know anything about multi-threading vs. multi-processing.
We should continue to export what we do, as some information are used by our new multiprocessing, to properly setup the environment.

**Comment by dlibenzi** [comment]
BTW, I still have to handle the 1 core case, which we should support.
I will be doing this in the PR, and not waiting for a follow up.

File: torch_xla_py/xla_multiprocessing.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+  for worker in config.split('|'):
+    parts = worker.split(';')
+    if len(parts) != 3:
+      raise ValueError('Bad worker syntax: {}'.format(worker))
+    parsed.append((parts[0], int(parts[1]), parts[2]))
+  return parsed
+
+
+def _get_devices_per_worker():
+  return 8
```
Comment (taylanbil) [inline torch_xla_py/xla_multiprocessing.py]: this reminds me, did we ever figure out if we can do multiprocessing on a single machine?

File: test/test_train_mp_mnist.py [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+    accuracy = test_loop_fn(para_loader.per_device_loader(device))
+    if FLAGS.metrics_debug:
+      print(torch_xla._XLAC._xla_metrics_report())
+
+  return accuracy
+
+
+def _mp_fn(index, flags):
+  global FLAGS
+  FLAGS = flags
```
Comment (taylanbil) [inline test/test_train_mp_mnist.py]: what is the point of these 2 lines?

File: test/test_train_mp_mnist.py [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+  torch.set_default_tensor_type('torch.FloatTensor')
```
Comment (taylanbil) [inline test/test_train_mp_mnist.py]: I don't think I do this in fairseq, should I? Is it a requirement?

File: test/test_train_mp_mnist.py [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+  accuracy = train_mnist()
```
Comment (taylanbil) [inline test/test_train_mp_mnist.py]: are these accuracies synced b/w multiprocesses? If not, shouldn't we sync first then compare against target?

File: test/test_train_mp_mnist.py [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+  if FLAGS.tidy and os.path.isdir(FLAGS.datadir):
+    shutil.rmtree(FLAGS.datadir)
+  if accuracy < FLAGS.target_accuracy:
+    print('Accuracy {} is below target {}'.format(accuracy,
```
Comment (taylanbil) [inline test/test_train_mp_mnist.py]: AFAICT these accuracies aren't synced at this point. So one subprocess can be below and another can be above. I feel like we should sync accuracies first and then check against target only once.

File: test/test_train_mp_mnist.py (hunk 7 above)
Comment (taylanbil) [inline test/test_train_mp_mnist.py]: what do these 2 lines do?

File: torch_xla_py/xla_env_vars.py (hunk 4 above)
Comment (taylanbil) [reply torch_xla_py/xla_env_vars.py]: shall we stick to pytorch lingo? I think they call this "RANK"

File: torch_xla_py/xla_multiprocessing.py (hunk 6 above)
Comment (taylanbil) [inline torch_xla_py/xla_multiprocessing.py]: this reminds me, did we ever figure out if we can do multiprocessing on a single machine, per core?

File: test/test_train_mp_mnist.py (hunk 8 above)
Comment (taylanbil) [inline test/test_train_mp_mnist.py]: I don't think I do this in fairseq, should I?

File: torch_xla_py/xla_env_vars.py (hunk 4 above)
Comment (dlibenzi) [reply torch_xla_py/xla_env_vars.py]: I have no strong opinion, but to me "ordinal" sounds better.
But isn't RANK the world size?
Anyway, this involves changes in xla dist as well, so might be topic for a different PR IMHO.

File: torch_xla_py/xla_multiprocessing.py (hunk 6 above)
Comment (dlibenzi) [reply torch_xla_py/xla_multiprocessing.py]: I change this to use:
```Python
return int(os.environ.get(xenv.TPU_NUM_DEVICES, '8'))
```
As we already have that configurable on the C++ side.

Comment (dlibenzi) [reply torch_xla_py/xla_multiprocessing.py]: Multiprocessing must work for a single machine and for POD/Slice.

File: test/test_train_mp_mnist.py (hunk 7 above)
Comment (dlibenzi) [reply test/test_train_mp_mnist.py]: It is in the PR comment.
Essentially pytorch forces its multiprocessing "vector" to be "spawn".
This means that there is no `fork()`, but instead a new process is created, and the Python state pickled and sent to the new process.
And `multiprocessing.spawn()` only pickles whatever is passed to the target function.
This means that any changes to the global state before the `spawn()` call, will not be reflected into the child processes.

File: test/test_train_mp_mnist.py (hunk 8 above)
Comment (dlibenzi) [reply test/test_train_mp_mnist.py]: Honestly, I think float is the default, so should not be necessary.

File: test/test_train_mp_mnist.py (hunk 9 above)
Comment (dlibenzi) [reply test/test_train_mp_mnist.py]: With multiprocessing there is hard separation between the different processes.
So, as you notice, we run the full validation set on each core.
We could have the parent/master process to create a queue, and pass it to the child processes, and receive accuracies back, but I am not sure it is worth it ATM.
We could do that later, though adds complexity to example models which IMHO should be kept as lean as possible.

File: test/test_train_mp_mnist.py (hunk 10 above)
Comment (dlibenzi) [reply test/test_train_mp_mnist.py]: See previous answer.

**Comment by dlibenzi** [comment]
I am not sure why, but I get duplicated comments from Taylan review :|

**Comment by dlibenzi** [comment]
BTW, I was planning to create a `test_train_mp_imagenet.py` to test performance ... but I should maybe be waiting for @zcain117 changes to it...

File: test/test_train_mp_mnist.py (hunk 3 above)
Comment (zcain117) [reply test/test_train_mp_mnist.py]: Does that mean that each device would have a different accuracy? That would mean some devices could pass the test and some could fail.

Or do we sync the model weights before computing accuracy, and each device runs over the entire test set, so each device should have the same accuracy?

**Comment by zcain117** [comment]
> BTW, I was planning to create a `test_train_mp_imagenet.py` to test performance ... but I should maybe be waiting for @zcain117 changes to it...

Definitely best to wait, especially given the imagenet regression. I haven't been able to find the culprit yet - it potentially could be in test_train_imagenet.py

File: test/test_train_mp_mnist.py (hunk 3 above)
Comment (dlibenzi) [reply test/test_train_mp_mnist.py]: Since we are running the whole validation set for each device, they should be getting same accuracy or very close.

**Comment by taylanbil** [comment]
I started a review than didn't finish, and forgot tab. So I restarted, and fginished that one. It picked up on the previous comments, so you see duplicates. Come on github!

**Comment by dlibenzi** [comment]
> > BTW, I was planning to create a `test_train_mp_imagenet.py` to test performance ... but I should maybe be waiting for @zcain117 changes to it...
> 
> Definitely best to wait, especially given the imagenet regression. I haven't been able to find the culprit yet - it potentially could be in test_train_imagenet.py

So I created one because I need to test it.
Looking at your LR changes, they do not look too invasive, so we can apply them to the MP version as well later (or earlier, if your PR gets in before).

File: torch_xla_py/xla_env_vars.py (hunk 4 above)
Comment (taylanbil) [reply torch_xla_py/xla_env_vars.py]: IIRC RANK is the index of a particular machine, world size is total num of devices (or processes, not sure). I'm fine either way, ORDINAL it is.

File: torch_xla_py/xla_multiprocessing.py (hunk 6 above)
Comment (taylanbil) [reply torch_xla_py/xla_multiprocessing.py]: I meant to ask single machine, per core. So, replace dataparallel threading w/ multiprocessing, essentially.

File: test/test_train_mp_mnist.py (hunk 9 above)
Comment (taylanbil) [reply test/test_train_mp_mnist.py]: Ah, so accuracies should be the same at all times then. SGTM.

File: torch_xla_py/xla_multiprocessing.py (hunk 6 above)
Comment (dlibenzi) [reply torch_xla_py/xla_multiprocessing.py]: One core is allowed. Or all cores. TPU does not allow to create ad-hoc meshes.
We could enable N cores (N < ALL), with replication disabled.
Inference would be fine, but as soon as you train (and hence CrossReplicaSum) you get hanging.

File: test/test_train_mp_mnist.py (hunk 7 above)
Comment (taylanbil) [reply test/test_train_mp_mnist.py]: I see. This looks weird, can we then pass FLAGS around as an arg?

```python
def train_mnist(args):
  ... (s/FLAGS/args/g)
```

wdyt?

File: torch_xla_py/xla_multiprocessing.py (hunk 6 above)
Comment (taylanbil) [reply torch_xla_py/xla_multiprocessing.py]: yep, so it's the same as before. I was asking because at some point we were going to look into if we could replace multithreading w/ multiprocessing (as that allows faster data loading apparently).

File: test/test_train_mp_mnist.py (hunk 7 above)
Comment (dlibenzi) [reply test/test_train_mp_mnist.py]: The convention for globals is all uppercase.
I am not sure what you mean with passing around?
Check the documentation of Python multiprocessing WRT the spawn way of creating processes.
This is what pytorch forces into, since I think they are not liking fork().

File: torch_xla_py/xla_multiprocessing.py (hunk 6 above)
Comment (dlibenzi) [reply torch_xla_py/xla_multiprocessing.py]: Yes, the replication constraints are not changed as far as TPU mesh goes.

File: test/test_train_mp_mnist.py (hunk 7 above)
Comment (taylanbil) [reply test/test_train_mp_mnist.py]: if I understand correctly, the issue here is, `train_mnist` must have access to the FLAGS object, but spawned subprocesses drop it. However, you're allowed to carry things to the subprocess thru the form of function args, so you do

```
global FLAGS
FLAGS = flags
```

to re-make the global FLAGS object in every subprocess, so `train_mnist` doesn't error.

I'm suggesting to add an argument to train_mnist, and drop the dependency to the global variable. That's pretty much what fairseq does and I think it would work here too, no?

Comment (dlibenzi) [reply test/test_train_mp_mnist.py]: What is happening is that the main process starts, and goes through the global scope code, which in our cases populates the global FLAGS variable.
That flags variable will not be populated in the child processes since no arguments will be passed. The new process will get a special argument for the pickled data.
Every piece of state which is modified in the master/parent process, which you want carried over to the child, must be a target function argument.

As far as replacing FLAGS with args, maybe not for this PR.
The is some a bit tricky stuff in test_utils.py that needs to be figured out.

**Comment by dlibenzi** [comment]
This is 15..12% boost on resnet50 on TPU v3.

**Comment by taylanbil** [comment]
Thanks to Daniel I now realize I misunderstood this PRs purpose. This
effectively allows multiprocessing on a single machine, that is going to be
a big boost for fairseq!

**Comment by dlibenzi** [comment]
Well, I am not sure about that. The boost should be bigger for models doing quite a bit of Python work.
Resnet50 gave 15..20% boost to me on TPU v3 (batch 128, 8 cores).

**Comment by dlibenzi** [comment]
@taylanbil @zcain117 are you OK with this or have more comments?
